File: Data Structures & Algorithms/merge-intervals/submission-2.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0],b[0]));

        List<int[]> ret = new LinkedList<>();
        ret.add(intervals[0]);

        for(int i = 0; i < intervals.length; i++){
            int lastEnd = ret.getLast()[1];

            if(intervals[i][0] <= lastEnd){
                ret.getLast()[1] = Math.max(intervals[i][1],lastEnd);
            }
            else{
                ret.add(intervals[i]);
            }
        }

        int[][] retArr = new int[ret.size()][2];

        for(int i = 0; i < retArr.length; i++){
            retArr[i] = ret.removeFirst();
        }

        return retArr;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        counts = [0] * 3
        for char in s:
            if char == '(':
                counts[0]+=1
            if char == ')':
                counts[1]+=1
            if char == '*':
                counts[2]+=1
        if abs(counts[0]-counts[1]) > counts[2]:
            return False
        
        lo = 0
        hi = 0
        for char in s:
            if char == ')':
                lo-=1
                hi-=1 
            elif char == '(':
                lo+=1
                hi+=1
            else:
                lo -= 1
                hi += 1
            if hi < 0:
                    return False
            lo = max(0,lo)
        return lo == 0
            

        